lib/using_restricted_array.rb

```diff
@@ -6,56 +6,126 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
-def length(array)
-  raise NotImplementedError
-end
+# Time complexity: O(n)
+# Space complexity: O(1)
+def length(array, length, value)
+	index = 0
+		
+	restricted_array.length.times do |index|
+		if restricted_array[index] == value
+			return true
+		end 
+	end 
+
+	return false
+end 
+
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  i = 0
+  while restricted_array != nil
+    puts restricted_array[i]
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(1)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
```

Comment:
Since you have to (worst case) search the entire length of the array, the time complexity is O(n)